### PR TITLE
Converted core fields to Typescript

### PR DESCRIPTION
**Pull request by heath-freenome**
### Reasons for making this change

- Updated the `@rjsf/utils` types and functions based on necessary changes needed for the conversion as follows:
  - In `types.ts` some improvements and new types were discovered as follows:
    - Modified the `RJSFSchema` to add the non-standard `enumNames` property onto the `JSONSchema7` to support the `Field`s
    - Added a new `EnumOptionsType` required for some of the fields
    - Added a new `UnsupportedFieldProps` type for the new `UnsupportedFieldTemplate` on `TemplatesType`
    - Removed the `fields` property on `FieldTemplateProps` since that data can be easily obtained from the `registry`
    - Added the `rawErrors` property to `ArrayFieldTemplateProps` since it was needed `ArrayField`
    - Added the missing `T` generic on a few uses of `IdSchema` properties on a few types
    - Updated the `WidgetProps` type to combine the `options` property with an additional, optional `enumOptions` property as needed by widgets
    - Updated the `UIOptionsBaseType` to have it combined with the `TemplatesType` making it require the `T` and `F` generics as well as adding the following new properties:
      - `classNames` was added as a breaking change (it used to simply be `classNames` in `UiSchema` but now it will be `ui:classNames` or `ui:options: { classNames }`
        - This makes `classNames` consistent with all the other ui options
      - `orderable` and `removable` were added to support the `ArrayField`
      - `widget` was added to support many of the `Field`s
    - Updated `UIOptionsType` to also take the `T` and `F` generics
    - Updated `SchemaUtilsType.retrieveSchema()` to make the `formData` parameter optional, like it already is on the wrapped function
  - Updated the `getUiOptions()` function to add the `T` and `F` generics to the return value type
  - Updated `getWidget()` to make the `widget` property optional, updating the tests accordingly
  - Updated `isCustomWidget()` to make `uiSchema` default to an empty object if missing, updating the tests accordingly
  - Updated `optionsList` to remove the `enumNames` hack and to return the new `EnumOptionsType[] | undefined` type
- Updated `FieldTemplate` for `antd` to remove the commented out `fields` variable since it was removed from the `FieldTemplateProps`
- Refactored `ArrayField.js` to `ArrayField.tsx` converting things to typescript
  - Made the `FieldProps` use the `T[]` generic based on the `ArrayField`'s `T` generic since `formData` should always be an array
  - There were many places where type safety checks were required along with custom casts to deal with `T` vs `T[]` issues
  - Exported `ArrayField` cast as a `Field` to allow it to be used properly in the `index.ts`
- Refactored `BooleanField.js` to `BooleanField.tsx` converting things to typescript
- Refactored `MultiSchemaField.js` to `MultiSchemaField.tsx` converting things to typescript
- Refactored `NullField.js` to `NullField.tsx` converting things to typescript
- Refactored `NumberField.js` to `NumberField.tsx` converting things to typescript as well as making it a stateless functional component with `useState()` instead of a class
- Refactored `ObjectField.js` to `ObjectField.tsx` converting things to typescript
- Refactored `SchemaField.js` to `SchemaField.tsx` converting things to typescript
  - Updated the code to deal with the `classNames` now coming from `ui:` options, including updating tests accordingly adding a new one for the new case inside of `ui:options`
- Refactored `StringField.js` to `StringField.tsx` converting things to typescript
- Refactored `fields/index.js` to `fields/index.ts` adding the proper type for the returned object
- Updated `ArrayFieldTemplate` to add the missing `T` and `F` generics required for `getUiOptions()`
- Updated `ObjectFieldTemplate` to add the `T` and `F` generics onto `getUiOptions()`
- Refactored `fields/UnsupportedField.js` to `templates/UnsupportedField.tsx` converting things to typescript
  - Added this to `templates/index.ts` as `UnsupportedFieldTemplate`
  - Also updated tests to use `templates.UnsupportedFieldTemplate` instead of `fields.UnsupportedField`
- Updated `getDefaultRegistry()` to remove the cast on `fields` since they are now properly typed
- Updated the test snapshots in the themes to deal with minor formatting updates due to these changes
- Renamed the `ArrayFieldTemplate.js` file in the `antd` theme to `index.js` to match all the other files
- Updated many of the `ObjectFieldTemplate`s to also recognize and support rendering the `description` from `uiOptions`

### Checklist

* [ ] **I'm updating documentation**
  - [ ] I've [checked the rendering](https://react-jsonschema-form.readthedocs.io/en/latest/#contributing) of the Markdown text I've added
* [x] **I'm adding or updating code**
  - [x] I've added and/or updated tests. I've run `npm run test:update` to update snapshots, if needed.
  - [ ] I've updated [docs](https://react-jsonschema-form.readthedocs.io/) if needed
  - [ ] I've updated the [changelog](https://github.com/rjsf-team/react-jsonschema-form/blob/master/CHANGELOG.md) with a description of the PR
* [ ] **I'm adding a new feature**
  - [ ] I've updated the playground with an example use of the feature

### CODE REVIEW

File: packages/core/src/components/fields/SchemaField.tsx [hunk 1]
```diff
@@ -35,19 +70,25 @@ function getFieldComponent(schema, uiSchema, idSchema, fields) {
   return componentName in fields
     ? fields[componentName]
     : () => {
-        const { UnsupportedField } = fields;
+        // cast this as a Partial to avoid issues
```
Comment (heath-freenome) [inline packages/core/src/components/fields/SchemaField.tsx:73]: Comment no longer needed.

File: packages/core/src/components/fields/ObjectField.tsx [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+
+  /** Handles the adding of a new additional property on the given `schema`. Calls the `onChange` callback once the new
+   * default data for that field has been added to the formData.
+   *
+   * @param schema - The schema element to which the new property is being added
+   */
+  handleAddClick = (schema: RJSFSchema) => () => {
+    if (typeof schema.additionalProperties !== "object") {
+      return;
+    }
```
Comment (heath-freenome) [inline packages/core/src/components/fields/ObjectField.tsx:189]: Use `lodash.isObject()` instead?

File: packages/core/src/components/fields/BooleanField.tsx [hunk 3]
```diff
@@ -22,17 +32,24 @@ function BooleanField(props) {
...
+        .map((option: RJSFSchemaDefinition) => {
+          if (isObject(option)) {
+            return {
+              ...option,
+              title: option.title || (option.const === true ? "Yes" : "No"),
+            };
+          }
+          return undefined;
+        })
+        .filter((o) => o) as RJSFSchemaDefinition[], // cast away the error that typescript can't grok is fixed
```
Comment (nickgros) [inline packages/core/src/components/fields/BooleanField.tsx:52]:
```suggestion
        .filter((o): o is RJSFSchemaDefinition => o) as // cast away the error that typescript can't grok is fixed
```

File: packages/core/src/components/fields/ArrayField.tsx [hunk 4]
```diff
@@ -0,0 +1,859 @@
...
+   */
+  onDropIndexClick = (index: number) => {
+    return (event: MouseEvent) => {
+      if (event) {
+        event.preventDefault();
+      }
+      const { onChange, errorSchema } = this.props;
+      const { keyedFormData } = this.state;
+      // refs #195: revalidate to ensure properly reindexing errors
+      let newErrorSchema: ErrorSchema<T>;
```
Comment (heath-freenome) [inline packages/core/src/components/fields/ArrayField.tsx:262]: Can this be a <T[]> now?

File: packages/core/src/components/fields/ArrayField.tsx [hunk 5]
```diff
@@ -0,0 +1,859 @@
...
+   * @param newIndex - The index to where the item is to be moved
+   */
+  onReorderClick = (index: number, newIndex: number) => {
+    return (event: React.MouseEvent<HTMLButtonElement>) => {
+      if (event) {
+        event.preventDefault();
+        event.currentTarget.blur();
+      }
+      const { onChange, errorSchema } = this.props;
+      let newErrorSchema: ErrorSchema<T>;
```
Comment (heath-freenome) [inline packages/core/src/components/fields/ArrayField.tsx:303]: Can this be a `<T[]>` now?

File: packages/core/src/components/fields/ArrayField.tsx [hunk 6]
```diff
@@ -0,0 +1,859 @@
...
+    const {
+      schema,
+      uiSchema = {},
+      errorSchema,
+      idSchema,
+      name,
+      required,
+      disabled,
+      readonly,
+      autofocus,
```
Comment (heath-freenome) [inline packages/core/src/components/fields/ArrayField.tsx:421]: Default these to `false`

File: packages/core/src/components/fields/ArrayField.tsx [hunk 7]
```diff
@@ -0,0 +1,859 @@
...
+    const _schemaItems = isObject(schema.items)
+      ? (schema.items as RJSFSchema)
+      : {};
+    const itemsSchema = schemaUtils.retrieveSchema(_schemaItems);
+    const formData = keyedToPlainFormData(this.state.keyedFormData);
+    const arrayProps: ArrayFieldTemplateProps<T[], F> = {
+      canAdd: this.canAddItem(formData),
+      items: keyedFormData.map((keyedItem, index) => {
+        const { key, item } = keyedItem;
+        // While we are actually dealing with a single item of type T, the
```
Comment (heath-freenome) [inline packages/core/src/components/fields/ArrayField.tsx:443]: Finish comment

File: packages/core/src/components/fields/ArrayField.tsx [hunk 8]
```diff
@@ -0,0 +1,859 @@
...
+        const itemCast = item as unknown as T[];
+        const itemSchema = schemaUtils.retrieveSchema(_schemaItems, itemCast);
+        const itemErrorSchema = errorSchema
+          ? (errorSchema[index] as ErrorSchema<T[]>)
+          : undefined;
+        const itemIdPrefix = idSchema.$id + idSeparator + index;
+        const itemIdSchema = schemaUtils.toIdSchema(
+          itemSchema,
+          itemIdPrefix,
+          item as unknown as T[],
```
Comment (heath-freenome) [inline packages/core/src/components/fields/ArrayField.tsx:453]: Use `itemCast`

File: packages/core/src/components/fields/ArrayField.tsx [hunk 9]
```diff
@@ -0,0 +1,859 @@
...
+   */
+  renderFiles() {
+    const {
+      schema,
+      uiSchema,
+      idSchema,
+      name,
+      disabled,
+      readonly,
+      autofocus,
```
Comment (heath-freenome) [inline packages/core/src/components/fields/ArrayField.tsx:609]: Default to false and add required

File: packages/core/src/components/fields/ArrayField.tsx [hunk 10]
```diff
@@ -0,0 +1,859 @@
...
+    const {
+      schema,
+      idSchema,
+      uiSchema,
+      formData: items = [],
+      disabled,
+      readonly,
+      required,
+      placeholder,
+      autofocus,
```
Comment (heath-freenome) [inline packages/core/src/components/fields/ArrayField.tsx:559]: Default to false (ignore placeholder)

File: packages/core/src/components/fields/ArrayField.tsx [hunk 11]
```diff
@@ -0,0 +1,859 @@
...
+      formData = [],
+      errorSchema,
+      idPrefix,
+      idSeparator = "_",
+      idSchema,
+      name,
+      required,
+      disabled,
+      readonly,
+      autofocus,
```
Comment (heath-freenome) [inline packages/core/src/components/fields/ArrayField.tsx:658]: Default to false

File: packages/core/src/components/fields/ArrayField.tsx [hunk 12]
```diff
@@ -0,0 +1,859 @@
...
+      onAddIndexClick: this.onAddIndexClick,
+      onDropIndexClick: this.onDropIndexClick,
+      onReorderClick: this.onReorderClick,
+      readonly,
+      registry,
+    };
+  }
+}
+
+export default ArrayField as unknown as Field;
```
Comment (heath-freenome) [inline packages/core/src/components/fields/ArrayField.tsx:859]: Describe why this is cast

File: packages/core/src/components/fields/BooleanField.tsx (hunk 3 above)
Comment (heath-freenome) [reply packages/core/src/components/fields/BooleanField.tsx:52]: Hmmm, not happy with that:
Screen Shot 2022-08-12 at 9 17 17 AM

File: packages/core/src/components/fields/ArrayField.tsx (hunk 4 above)
Comment (heath-freenome) [reply packages/core/src/components/fields/ArrayField.tsx:262]: Nope:
Screen Shot 2022-08-12 at 9 19 01 AM

**Review by nickgros — APPROVED** [review]

**Pull request merged**